### PR TITLE
Improve verification of authorization in management of permissions

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java [hunk 1]
```diff
@@ -42,9 +43,9 @@ public static void checkProjectAdminUserByComponentKey(UserSession userSession,
     }
   }
 
-  public static void checkProjectAdminUserByComponentUuid(UserSession userSession, @Nullable String componentUuid) {
+  public static void checkAdministrationPermission(UserSession userSession, Optional<ProjectId> projectId) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java]: MINOR Specify a "ProjectId" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java [hunk 2]
```diff
@@ -35,6 +35,9 @@
   @CheckForNull
   Integer getUserId();
 
+  /**
+   * @deprecated does not support organizations as group names are not unique
+   */
   Set<String> getUserGroups();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java]: MAJOR Add the missing @Deprecated annotation. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* MAJOR 2 major
* MINOR 2 minor
* INFO 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [ThreadLocalUserSession.java#L76](https://github.com/SonarSource/sonarqube/blob/caad5d24407f708723ed219595ed15129801f8ce/server/sonar-server/src/main/java/org/sonar/server/user/ThreadLocalUserSession.java#L76): Remove this use of "getUserGroups"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java [hunk 3]
```diff
@@ -42,9 +43,26 @@ public static void checkProjectAdminUserByComponentKey(UserSession userSession,
     }
   }
 
-  public static void checkProjectAdminUserByComponentUuid(UserSession userSession, @Nullable String componentUuid) {
+  /**
+   * Checks that user is administrator of the specified project, or of the specified organization if project is not
+   * defined.
+   * @throws org.sonar.server.exceptions.ForbiddenException if user is not administrator
+   */
+  public static void checkAdministratorOfProjectOrOrganization(UserSession userSession, String organizationUuid, Optional<ProjectId> projectId) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java]: MINOR Specify a "ProjectId" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java [hunk 4]
```diff
@@ -42,9 +43,26 @@ public static void checkProjectAdminUserByComponentKey(UserSession userSession,
...
+      userSession.checkOrganizationPermission(organizationUuid, GlobalPermissions.SYSTEM_ADMIN);
+    }
+  }
+
+  /**
+   * Checks that user is administrator of the specified project, or of system if project is not
+   * defined.
+   * @throws org.sonar.server.exceptions.ForbiddenException if user is not administrator
+   */
+  public static void checkAdministratorOfProject(UserSession userSession, Optional<ProjectId> projectId) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java]: MINOR Specify a "ProjectId" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java [hunk 5]
```diff
@@ -35,6 +35,10 @@
   @CheckForNull
   Integer getUserId();
 
+  /**
+   * @deprecated does not support organizations as group names are not unique
+   */
+  @Deprecated
   Set<String> getUserGroups();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* MAJOR 1 major
* MINOR 3 minor
* INFO 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [ThreadLocalUserSession.java#L76](https://github.com/SonarSource/sonarqube/blob/5129a88212167b28b1dcfbf13e44b889745b63d9/server/sonar-server/src/main/java/org/sonar/server/user/ThreadLocalUserSession.java#L76): Remove this use of "getUserGroups"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/permission/ProjectId.java [hunk 6]
```diff
@@ -25,21 +25,23 @@
...
- * Reference to a project by its db id or uuid. Temporarily
- * as long as permissions do not use only uuids.
+ * Reference to a project by its db id or uuid. The field "id" should
+ * be removed as soon as backend is fully based on uuids.
+ *
+ * @see org.sonar.server.permission.ws.WsProjectRef
  */
 @Immutable
-public class ProjectRef {
+public class ProjectId {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ProjectId.java]: the id is supposed to be a legacy thing and be dropped asap
still, the name of the class is now Project**Id**.
the `ProjectRef` name seems more logical to me (because it is neither UUID nor ID).
but ok

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/SearchProjectPermissionsAction.java [hunk 7]
```diff
@@ -60,13 +60,16 @@
   private final I18n i18n;
   private final ResourceTypes resourceTypes;
   private final SearchProjectPermissionsDataLoader dataLoader;
+  private final PermissionWsSupport wsSupport;
 
-  public SearchProjectPermissionsAction(DbClient dbClient, UserSession userSession, I18n i18n, ResourceTypes resourceTypes, SearchProjectPermissionsDataLoader dataLoader) {
+  public SearchProjectPermissionsAction(DbClient dbClient, UserSession userSession, I18n i18n, ResourceTypes resourceTypes,
+                                        SearchProjectPermissionsDataLoader dataLoader, PermissionWsSupport wsSupport) {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/SearchProjectPermissionsAction.java]: bad indent

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java (hunk 1 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java]: This method name makes me wonder what "administration permission" is.

Checking the 'admin' (value in DB) is referred to as "admin" pretty much everywhere and especially in the other methods of this class.

Why not simply call this method `checkProjectAdminUser` or `checkProjectAdmin`? (I prefer the later but less consistant with other methods of the class)

File: server/sonar-server/src/test/java/org/sonar/server/permission/GroupPermissionChangerTest.java [hunk 8]
```diff
@@ -50,7 +48,7 @@
   public ExpectedException expectedException = ExpectedException.none();
 
   private UserSessionRule userSession = UserSessionRule.standalone();
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/permission/GroupPermissionChangerTest.java]: can't you remove the rule completely?

File: sonar-db/src/main/java/org/sonar/db/permission/AuthorizationMapper.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.Set;
+import org.apache.ibatis.annotations.Param;
+
+/**
+ * @see AuthorizationDao
+ */
+public interface AuthorizationMapper {
+
+  Set<String> selectOrganizationPermissions(@Param("organizationUuid") String organizationUuid,
+    @Param("userId") long userId);
```
Comment (sns-seb) [inline sonar-db/src/main/java/org/sonar/db/permission/AuthorizationMapper.java]: line return not required

File: sonar-db/src/main/java/org/sonar/db/permission/PermissionMapper.java [hunk 10]
```diff
@@ -26,8 +26,8 @@
 /**
  * Only the requests involving both user and group permissions.
```
Comment (sns-seb) [inline sonar-db/src/main/java/org/sonar/db/permission/PermissionMapper.java]: "request" => "queries"?

File: sonar-db/src/main/resources/org/sonar/db/permission/AuthorizationMapper.xml [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+<mapper namespace="org.sonar.db.permission.AuthorizationMapper">
+
+  <select id="selectOrganizationPermissions" parameterType="map" resultType="string">
+    select gr.role
+    from group_roles gr
+    where
+    gr.organization_uuid=#{organizationUuid,jdbcType=VARCHAR} and
+    gr.resource_id is null and
+    (gr.group_id is null or gr.group_id in (select gu.group_id from groups_users gu where gu.user_id=#{userId,jdbcType=BIGINT}))
```
Comment (sns-seb) [inline sonar-db/src/main/resources/org/sonar/db/permission/AuthorizationMapper.xml]: I think a join or outer join or another select in the union should be preferred to a subselect, for efficiency

I think UNION is quite a good option as many SGDB now run each select in // under the hood (as @eric-duquesnoy told me yesterday)

File: sonar-db/src/main/resources/org/sonar/db/permission/AuthorizationMapper.xml [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    gr.resource_id is null and
+    gr.group_id is null
+  </select>
+
+  <select id="selectRootComponentPermissions" parameterType="map" resultType="string">
+    select gr.role
+    from group_roles gr
+    where
+    gr.resource_id=#{rootComponentId,jdbcType=BIGINT} and
+    (gr.group_id is null or gr.group_id in (select gu.group_id from groups_users gu where gu.user_id=#{userId,jdbcType=BIGINT}))
```
Comment (sns-seb) [inline sonar-db/src/main/resources/org/sonar/db/permission/AuthorizationMapper.xml]: same comment as above, try and avoid sub selects

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* MAJOR 1 major
* MINOR 3 minor
* INFO 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [ThreadLocalUserSession.java#L76](https://github.com/SonarSource/sonarqube/blob/ad25bdbae8ad3e581186c983ddae8752e8a53b17/server/sonar-server/src/main/java/org/sonar/server/user/ThreadLocalUserSession.java#L76): Remove this use of "getUserGroups"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java [hunk 13]
```diff
@@ -131,6 +133,30 @@ public boolean isRoot() {
   }
 
   @Override
+  public boolean hasOrganizationPermission(String organizationUuid, String permission) {
+    if (permissionsByOrganizationUuid == null) {
+      permissionsByOrganizationUuid = HashMultimap.create();
+    }
+    Set<String> permissions;
+    if (permissionsByOrganizationUuid.containsKey(organizationUuid)) {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: you can avoid calling both `containsKey` and `get` (they essentially do the same thing):
```java
Set<String> permissions = permissionsByOrganizationUuid.get(organizationUuid);
if (permissions == null) {
   permissions = loadOrganizationPermissions(organizationUuid);
   permissionsByOrganizationUuid.putAll(organizationUuid, permissions);
}
return permissions.contains(permission);
```

File: server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java (hunk 5 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java]: I don't see any replacement provided in the PR.
This might be caused by the main consumer of this method is the issues domain (and the ES index) and the change to fix it is quite big and should be made in a dedicated PR, right?

File: server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java [hunk 14]
```diff
@@ -80,6 +84,18 @@
   boolean hasPermission(String globalPermission);
 
   /**
+   * Returns {@code true} if the permission is granted on the organization, else {@code false}.
+   * Root status is not verified, so the method may return {@code false} even for root users.
+   */
+  boolean hasOrganizationPermission(String organizationUuid, String permission);
+
+  /**
+   * Ensures that user implies the specified organization permission,
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java]: "user implies the specified..." <- me no understand you English :p

File: server/sonar-server/src/test/java/org/sonar/server/permission/ws/AddGroupActionTest.java [hunk 15]
```diff
@@ -226,7 +245,8 @@ public void fail_when_project_uuid_and_project_key_are_provided() throws Excepti
   @Test
   public void adding_global_permission_fails_if_not_administrator_of_organization() throws Exception {
     GroupDto group = db.users().insertGroup(defaultOrganizationProvider.getDto(), "sonar-administrators");
-    userSession.login();
+    // user is administrator of another organization
+    userSession.login().addOrganizationPermission("anotherOrg", SYSTEM_ADMIN);
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/permission/ws/AddGroupActionTest.java]: you could use an override of method `loginAsOrganizationAdmin` here, you could then remove the comment

File: server/sonar-server/src/test/java/org/sonar/server/permission/ws/AddUserActionTest.java [hunk 16]
```diff
@@ -246,6 +247,10 @@ public void adding_project_permission_is_allowed_to_project_administrators() thr
   }
 
   private void loginAsAdmin() {
-    userSession.login().setGlobalPermissions(SYSTEM_ADMIN);
+    loginAsOrganizationAdmin(db.getDefaultOrganization());
+  }
+
+  private void loginAsOrganizationAdmin(OrganizationDto org) {
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/permission/ws/AddUserActionTest.java]: isn't a test missing making use of this method with an organization which is not the default one?

File: server/sonar-server/src/test/java/org/sonar/server/permission/ws/RemoveGroupActionTest.java [hunk 17]
```diff
@@ -316,7 +317,11 @@ public void removing_project_permission_is_allowed_to_project_administrators() t
   }
 
   private void loginAsAdmin() {
-    userSession.login().setGlobalPermissions(SYSTEM_ADMIN);
+    loginAsOrganizationAdmin(db.getDefaultOrganization());
+  }
+
+  private void loginAsOrganizationAdmin(OrganizationDto org) {
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/permission/ws/RemoveGroupActionTest.java]: I think a unit test is missing using this method with an organization which is not the default one

File: server/sonar-server/src/test/java/org/sonar/server/permission/ws/RemoveUserActionTest.java [hunk 18]
```diff
@@ -276,6 +277,10 @@ public void removing_project_permission_is_allowed_to_project_administrators() t
   }
 
   private void loginAsAdmin() {
-    userSession.login("admin").setGlobalPermissions(SYSTEM_ADMIN);
+    loginAsOrganizationAdmin(db.getDefaultOrganization());
+  }
+
+  private void loginAsOrganizationAdmin(OrganizationDto org) {
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/permission/ws/RemoveUserActionTest.java]: I think a unit test is missing using this method with an organization which is not the default one

File: server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java (hunk 14 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java]: bad reason: it's a copy paste of other public methods. I'm fixing that.

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java (hunk 13 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: nope, multimap returns an empty set if key is not present :-)

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java [hunk 19]
```diff
@@ -42,9 +43,28 @@ public static void checkProjectAdminUserByComponentKey(UserSession userSession,
...
+  }
+
+  /**
+   * Checks that user is administrator of the specified project, or of system if project is not
+   * defined.
+   * @throws org.sonar.server.exceptions.ForbiddenException if user is not administrator
+   * @deprecated does not support organizations. Replaced by {@link #checkProjectAdmin(UserSession, String, Optional)}
+   */
+  @Deprecated
+  public static void checkProjectAdmin(UserSession userSession, Optional<ProjectId> projectId) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java]: MINOR Specify a "ProjectId" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)
INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java [hunk 20]
```diff
@@ -42,9 +43,28 @@ public static void checkProjectAdminUserByComponentKey(UserSession userSession,
     }
   }
 
-  public static void checkProjectAdminUserByComponentUuid(UserSession userSession, @Nullable String componentUuid) {
+  /**
+   * Checks that user is administrator of the specified project, or of the specified organization if project is not
+   * defined.
+   * @throws org.sonar.server.exceptions.ForbiddenException if user is not administrator
+   */
+  public static void checkProjectAdmin(UserSession userSession, String organizationUuid, Optional<ProjectId> projectId) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java]: MINOR Specify a "ProjectId" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/UsersAction.java [hunk 21]
```diff
@@ -89,8 +89,8 @@ public void define(WebService.NewController context) {
   @Override
   public void handle(Request request, Response response) throws Exception {
     try (DbSession dbSession = dbClient.openSession(false)) {
-      Optional<ProjectRef> projectId = support.findProject(dbSession, request);
-      checkProjectAdminUserByComponentUuid(userSession, projectId.isPresent() ? projectId.get().getUuid() : null);
+      Optional<ProjectId> projectId = support.findProject(dbSession, request);
+      checkProjectAdmin(userSession, projectId);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/UsersAction.java]: MINOR Remove this use of "checkProjectAdmin"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/GroupsAction.java [hunk 22]
```diff
@@ -86,8 +86,8 @@ public void define(WebService.NewController context) {
   @Override
   public void handle(Request request, Response response) throws Exception {
     try (DbSession dbSession = dbClient.openSession(false)) {
-      Optional<ProjectRef> projectId = support.findProject(dbSession, request);
-      checkProjectAdminUserByComponentUuid(userSession, projectId.isPresent() ? projectId.get().getUuid() : null);
+      Optional<ProjectId> projectId = support.findProject(dbSession, request);
+      checkProjectAdmin(userSession, projectId);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/GroupsAction.java]: MINOR Remove this use of "checkProjectAdmin"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/SearchProjectPermissionsAction.java [hunk 23]
```diff
@@ -113,19 +118,15 @@ private static SearchProjectPermissionsWsRequest toSearchProjectPermissionsWsReq
...
+  private void checkAuthorized(DbSession dbSession, SearchProjectPermissionsWsRequest request) {
+    com.google.common.base.Optional<ProjectWsRef> projectRef = newOptionalWsProjectRef(request.getProjectId(), request.getProjectKey());
+    Optional<ProjectId> projectId;
+    if (projectRef.isPresent()) {
+      projectId = Optional.of(wsSupport.findProject(dbSession, projectRef.get()));
     } else {
-      checkGlobalAdminUser(userSession);
+      projectId = Optional.empty();
     }
+    checkProjectAdmin(userSession, projectId);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/SearchProjectPermissionsAction.java]: MINOR Remove this use of "checkProjectAdmin"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues
* MAJOR 1 major
* MINOR 6 minor
* INFO 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [ThreadLocalUserSession.java#L76](https://github.com/SonarSource/sonarqube/blob/6416b1d7e142567518f34504e85a51b75653ae99/server/sonar-server/src/main/java/org/sonar/server/user/ThreadLocalUserSession.java#L76): Remove this use of "getUserGroups"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/permission/UserPermissionChanger.java [hunk 24]
```diff
@@ -54,7 +48,12 @@ public boolean apply(DbSession dbSession, UserPermissionChange change) {
         break;
       case REMOVE:
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/UserPermissionChanger.java]: MAJOR Reduce this switch case number of lines from 8 to at most 5, for example by extracting code into methods. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1151)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues
* MAJOR 2 major
* MINOR 6 minor
* INFO 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [ThreadLocalUserSession.java#L76](https://github.com/SonarSource/sonarqube/blob/fca278e68ed5aafc0b8c9ec44a56d3147313af9a/server/sonar-server/src/main/java/org/sonar/server/user/ThreadLocalUserSession.java#L76): Remove this use of "getUserGroups"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-db/src/main/java/org/sonar/db/permission/AuthorizationMapper.java [hunk 25]
```diff
@@ -21,15 +21,23 @@
...
+
+  Set<String> selectOrganizationPermissions(@Param("organizationUuid") String organizationUuid, @Param("userId") long userId);
+
+  Set<String> selectOrganizationPermissionsOfAnonymous(@Param("organizationUuid") String organizationUuid);
+
+  Set<String> selectRootComponentPermissions(@Param("rootComponentId") long rootComponentId, @Param("userId") long userId);
+
+  Set<String> selectRootComponentPermissionsOfAnonymous(@Param("rootComponentId") long rootComponentId);
+
+  int countRemainingUserIdsWithGlobalPermissionIfExcludeGroup(@Param("organizationUuid") String organizationUuid, @Param("permission") String permission, @Param("excludedGroupId") long excludedGroupId);
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/AuthorizationMapper.java]: MINOR Split this 202 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: server/sonar-server/src/main/java/org/sonar/server/permission/UserPermissionChanger.java [hunk 26]
```diff
@@ -70,7 +69,7 @@ private boolean shouldSkipChange(DbSession dbSession, UserPermissionChange chang
 
   private void checkOtherAdminUsersExist(DbSession session, PermissionChange change) {
     if (GlobalPermissions.SYSTEM_ADMIN.equals(change.getPermission()) &&
-      !change.getProjectRef().isPresent() &&
+      !change.getProjectId().isPresent() &&
       dbClient.roleDao().countUserPermissions(session, change.getPermission(), null) <= 1) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/UserPermissionChanger.java]: MINOR Remove this use of "countUserPermissions"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-db/src/main/java/org/sonar/db/user/RoleDao.java [hunk 27]
```diff
@@ -49,6 +49,10 @@ public void deleteGroupRolesByGroupId(DbSession session, long groupId) {
     mapper(session).deleteGroupRolesByGroupId(groupId);
   }
 
+  /**
+   * @deprecated does not support organizations
+   */
+  @Deprecated
   public int countUserPermissions(DbSession session, String permission, @Nullable Long allGroupsExceptThisGroupId) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/user/RoleDao.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/AddUserAction.java [hunk 28]
```diff
@@ -65,18 +65,16 @@ public void define(NewController context) {
...
       String login = request.mandatoryParam(PARAM_LOGIN);
       UserDto user = dbClient.userDao().selectActiveUserByLogin(dbSession, login);
-      if (user == null) {
-        throw new NotFoundException(format("Could not find a user with login '%s'", login));
-      }
+      checkFound(user, "Could not find a user with login '%s'", login);
 
-      if (userIsNotYetMemberOf(dbSession, user.getId(), groupId)) {
+      if (!isMemberOf(dbSession, user, groupId)) {
         UserGroupDto membershipDto = new UserGroupDto().setGroupId(groupId.getId()).setUserId(user.getId());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/AddUserAction.java]: MAJOR NullPointerException might be thrown as 'user' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 14 issues
* MAJOR 3 major
* MINOR 8 minor
* INFO 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [ThreadLocalUserSession.java#L76](https://github.com/SonarSource/sonarqube/blob/c0bb0392d9a8d5c299d3b7a6d4dd14f20707f779/server/sonar-server/src/main/java/org/sonar/server/user/ThreadLocalUserSession.java#L76): Remove this use of "getUserGroups"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java [hunk 29]
```diff
@@ -35,16 +36,39 @@ public static void checkGlobalAdminUser(UserSession userSession) {
       .checkPermission(GlobalPermissions.SYSTEM_ADMIN);
   }
 
+  /**
+   * @deprecated does not support organizations. Replaced by {@link #checkProjectAdmin(UserSession, String, Optional)}
+   */
+  @Deprecated
   public static void checkProjectAdminUserByComponentKey(UserSession userSession, @Nullable String componentKey) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionPrivilegeChecker.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 15 issues
* MAJOR 3 major
* MINOR 8 minor
* INFO 4 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [ThreadLocalUserSession.java#L76](https://github.com/SonarSource/sonarqube/blob/5a0aa8d90f7d601a68364e817c01ee60036bcaf7/server/sonar-server/src/main/java/org/sonar/server/user/ThreadLocalUserSession.java#L76): Remove this use of "getUserGroups"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
